source/Mlos.Python/mlos/Optimizers/ExperimentDesigner/UtilityFunctions/MultiObjectiveProbabilityOfImprovementUtilityFunction.py

```diff
@@ -0,0 +1,211 @@
+        for _, prediction in multi_objective_predictions:
+            prediction_df = prediction.get_dataframe()
+            valid_predictions_index = valid_predictions_index.intersection(prediction_df.index)
+
+        # Let's make sure all predictions have a standard deviation available.
+        #
+        for _, objective_prediction in multi_objective_predictions:
+            std_dev_column_name = objective_prediction.add_standard_deviation_column()
+
+        batched_poi_df = self._batched_probability_of_improvement(
+            multi_objective_predictions=multi_objective_predictions,
+            valid_predictions_index=valid_predictions_index,
+            std_dev_column_name=std_dev_column_name
+        )
+
+        return batched_poi_df
+
+    @trace()
+    def _batched_probability_of_improvement(
+            self,
+            multi_objective_predictions: MultiObjectivePrediction,
+            valid_predictions_index: pd.Index,
+            std_dev_column_name: str
+    ):
+        """Generates a single large dataframe of monte carlo samples to send to ParetoFrontier for evaluation.
+
+        Profiling reveals that batching the query to the ParetoFrontier.is_dominated() function produces massive perf gains over the naive implementation.
+
+        A few additional optimizations are possible:
+            1) Maybe inline the self.create_monte_carlo_samples_df function. It decreases readability so should be justified.
+            2) First generate only a few samples and reject configs that are strictly dominated on the basis of that low-resolution
+               evidence. Then repeat the process at higher resolutions for the surviving subset.
+
+        :param multi_objective_predictions:
+        :param valid_predictions_index:
+        :return:
+        """
+        monte_carlo_samples_dfs = []
+
+        for config_idx in valid_predictions_index:
+            monte_carlo_samples_df = self.create_monte_carlo_samples_df(
+                multi_objective_predictions=multi_objective_predictions,
+                config_idx=config_idx,
+                std_dev_column_name=std_dev_column_name
+            )
+            monte_carlo_samples_df['config_idx'] = config_idx
+            monte_carlo_samples_dfs.append(monte_carlo_samples_df)
+
+        samples_for_all_configs_df = pd.concat(monte_carlo_samples_dfs, ignore_index=True, axis=0)
+        original_column_names = [column for column in samples_for_all_configs_df.columns if column != 'config_idx']
+        samples_for_all_configs_df['is_dominated'] = self.pareto_frontier.is_dominated(objectives_df=samples_for_all_configs_df[original_column_names])
+        poi_series = 1 - (samples_for_all_configs_df.groupby(by=["config_idx"])["is_dominated"].sum() / self.config.num_monte_carlo_samples)
+        poi_df = pd.DataFrame({'utility': poi_series}, index=poi_series.index)
+        return poi_df
+
+
+    @trace()
+    def _naive_poi(

```

Comment:
This function was superseded by _batched_probability_of_improvement. I left the naive implementation here as it's easier to understand.